flink-connectors/flink-connector-kafka-base/src/main/java/org/apache/flink/streaming/util/serialization/KeyedDeserializationSchema.java

```diff
@@ -45,6 +45,22 @@
 	 */
 	T deserialize(byte[] messageKey, byte[] message, String topic, int partition, long offset) throws IOException;
 
+	/**
+	 * Deserializes the byte message.
+	 *
+	 * @param messageKey the key as a byte array (null if no key has been set).
+	 * @param message The message, as a byte array (null if the message was empty or deleted).
+	 * @param partition The partition the message has originated from.
+	 * @param offset the offset of the message in the original source (for example the Kafka offset).
+	 * @param timestamp the timestamp of the consumer record
+	 * @param timestampType The timestamp type, could be NO_TIMESTAMP, CREATE_TIME or INGEST_TIME.
+	 *
+	 * @return The deserialized message as an object (null if the message cannot be deserialized).
+	 */
+	default T deserialize(byte[] messageKey, byte[] message, String topic, int partition, long offset, long timestamp, TimestampType timestampType) throws IOException {
```

Comment:
This would mean that, if users want to access timestamps, they would have to implement both deserialize methods, correct?

This feels a bit awkward. I can think of 2 possible approaches to address this:
1. Let the original non-timestamped deserialize method also be default, but have an  empty method. It's still not nice though, since now both deserialize methods are default, moreover the second deserialize method relies on the first one.

2. Instead of having a new deserialize method with the timestamp, we have an `setTimestamp` method that is only ever called in Kafka 0.10+, which signature is `void setTimestamp(T deserializedRecord, long timestamp)`. This might also match better what we've discussed in the JIRA ticket, about having separate interfaces for "deserialization" and "enrichment".

Reply:
I have a small problem with the second alternative. When I implement this interface in Scala, I do not want to use mutable objectS. I think the `setTimestamp` method is forcing to make the `T deserializedRecord` mutable (for at least the timestamp field)

I have no problem with the first alternative, but I think we are better of when we throw an exception with a message explaining that must implement/override one of the `deserialize`  methods.

Reply:
@FredTing Your first concern makes sense. Though, wouldn't this be a shortcoming also for our long-term solution, where we want to have separate interfaces for "deserialization" and "enrichment"?

```
I have no problem with the first alternative, but I think we are better of when we throw an exception with a message explaining that must implement/override one of the deserialize methods.
```

I'm not sure how this would work. Could you elaborate a bit more on this?

Reply:
@tzulitai Yes, this would also be an issue with our long-term solution, and we should look at that carefully. For instance, when messages are encrypted on Kafka we need them to be decrypted. How stuff is encrypted is determined by the producer (out of our control), but we should be able to decrypt these message. When for encryption some kafka-meta information is used we need this during decryption too.

> I'm not sure how this would work. Could you elaborate a bit more on this?

When a custom implementation of the `KeyedDeserializationSchema` is created, then at least one of the `deserialize` methods must be implement (why else create a custom implementation?). When both `deserialize` methods have a default implementation the compiler isn't complaining anymore if you forget to implement it. When it is an empty method, you won't get complains during runtime either, but you won't receive any record too. By throwing an exception with a clear message that you forgot to override either one of the `deserialize` methods, you get a error message during runtime.
When either one of the `deserialize` method is overridden the default with the exception isn't called any more. 
Since it's now an abstract method, al current implementations already override this `deserialize` method

Reply:
I'm still not very convinced that this is a good approach.

For example, things might be a lot cleaner and error-proof if we can just deprecate the non-timestamped `KeyedSerializationSchema` in favor of a new interface. Technically, we're trying to achieve the same thing, by things would perhaps be much understandable. 

@FredTing what do you think?
Perhaps @aljoscha would also want to chime in some thoughts here?

Reply:
@tzulitai We can remove the non-timestamped version of the `deserialize` method completely. It will break the interface and if you implemented this method it quite easy to migrate to the new method. I've only introduced the new `deserialize` method as default to make this change backwards compatible.
When backwards compatibility is required we could mark the non-timestamped version of the `deserialize` method as `@Deprecated` and give it a default implementation that throws a `NotImplementedException` or a `FlinkRuntimeException` with the message "you must implement the deserialize method when implementingt the KeyedDeserializationSchema interface"

Reply:
Yes, to unblock this I thing we can go with this approach, basically the schema becomes this:
```
@PublicEvolving
public interface KeyedDeserializationSchema<T> extends Serializable, ResultTypeQueryable<T> {

        @Deprecated
	default T deserialize(byte[] messageKey, byte[] message, String topic, int partition, long offset) {
            throw new RuntimeException("blammo");
        }

        default T deserialize(byte[] messageKey, byte[] message, String topic, int partition, long offset, long timestamp) {
            return deserialize(/* call the other method */);
        } 
	boolean isEndOfStream(T nextElement);
}
```

With this, if you have an existing implementation of `KeyedDeserializationSchema` it will continue to work without any changes. If you implement a new one you have to implement one of the methods, otherwise the exception is thrown. And all Flink code only calls the version that takes the timestamp.

What do you think?

Reply:
I would be ok with proceeding with the above.

Reply:
I've pushed the changes as we discussed above.